### PR TITLE
Check targetgroup health status during asg replacement

### PR DESCRIPTION
**Pull request by micarlise**
Targetgroup needs to be healthy in order for ALBs to route traffic to
them.  This patch requires nemesys to check for target group health
alongside elb and asg health status; all before moving to terminating
the old asg.

This patch should prevent an issue when replacing ASGs with only
targetgroups. The load balancer returns no unhealthy instances,
therefore nemesys proceeds to terminate the old ASG even though the new
ASG can't take new traffic yet.

The way we check is still leaky.  The passing condition is that there
are no unhealthy instances in either the target group or the load
balancer.  But 0 unhealthy instances and no healhty instances are not
synonymous.  In the future, we should probably switch to the
number of healthy instances across elbs and tgs needs to match the
desired capacity.  An additional health check would be to actually hit
the uri endpoint of the load balancer to make sure traffic is routable
before finally terminating the old asg.

### CODE REVIEW

File: src/api/asg/replace.js [hunk 1]
```diff
@@ -83,21 +83,40 @@ function _wait_for_health(region, new_asg_name, new_asg, old_asg) {
...
+								})
+							);
+							result = result.concat(tg_result
+								.filter(obj => {
+									return obj.TargetHealth.State != 'healthy';
+								}).map(instance => {
+									return instance.Target.Id;
+								})
+							);
+							return result;
```
Comment (torra) [inline src/api/asg/replace.js]: i like how you just look up both the ELB and TG health - seems simpler than having a conditional in here somewhere. this `result` array will have duplicates in the case that there is both an ELB and TG though, so the counts that get logged later could be off. i would dedupe it here so its not confusing in the logs

File: src/api/asg/replace.js [hunk 2]
```diff
@@ -83,21 +83,40 @@ function _wait_for_health(region, new_asg_name, new_asg, old_asg) {
 		return new Promise(function(resolve, reject){
 			function _check() {
 				AWSUtil.get_asg(AWSProvider.get_as(region), new_asg_name).then(function(asg){
-					let new_instance_ids = asg.Instances.map((instance) => instance.InstanceId);
-					let inst_elb_check_proms = asg.LoadBalancerNames.map(function(elb){
+					let inst_elb_check_proms = BB.all(asg.LoadBalancerNames.map(function(elb){
 						return AWSProvider.get_elb(region).describeInstanceHealthAsync({
-							LoadBalancerName: elb,
-							Instances: new_instance_ids.map( (inst_id) => ({InstanceId: inst_id}) )
```
Comment (torra) [inline src/api/asg/replace.js]: it looks like you are getting the health of all instances in the load balancer (and similarly for the target group below). i don't think that is what we want to do - i think we want to only check the health in the load balancer / target group of the instances that are part of the new ASG, hence this filter that was here before

**Comment by torra** [comment]
thinking about this a little more after reading your pr description again, i think it does make sense to look for these instances to explicitly be healthy, rather than checking that they are not unhealthy. so, once all the instances are healthy that are in the asg, it is time to start spinning down the old ones. lmk what you think, it doesnt seem like that would be a huge change

File: src/api/asg/replace.js (hunk 2 above)
Comment (micarlise) [reply src/api/asg/replace.js]: That instance map will have the instance ids for all instances in the ASG.  which would include both the load balancer and target groups.  I wasn't sure what would happen if we ask the elb for instance health status of an instance that isn't attached to the elb.  Will elbv1().describeInstancesHealthAsync() just ignore those instances?  I'll test that.

**Pull request merged**